XenAdmin/Diagnostics/Checks/HostMemoryPostUpgradeCheck.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System.Collections.Generic;
+using System.Web.Script.Serialization;
+using XenAdmin.Core;
+using XenAdmin.Diagnostics.Problems;
+using XenAdmin.Diagnostics.Problems.HostProblem;
+using XenAPI;
+
+namespace XenAdmin.Diagnostics.Checks
+{
+    class HostMemoryPostUpgradeCheck : HostPostLivenessCheck
+    {
+        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
+
+        private readonly Dictionary<string, string> installMethodConfig;
+
+        public HostMemoryPostUpgradeCheck(Host host, Dictionary<string, string> installMethodConfig)
+            : base(host)
+        {
+            this.installMethodConfig = installMethodConfig;
+        }
+        public override string Description => Messages.CHECKING_HOST_MEMORY_POST_UPGRADE_DESCRIPTION;
+
+        protected override Problem RunHostCheck()
+        {
+            if (TryGetDom0MemoryPostUpgrade(out var dom0MemoryPostUpgrade))
+            {
+                var currentDom0Memory = Host.dom0_memory();
+                // we know the Dom0 memory after the upgrade, check if it will greater then the current value
+                if (dom0MemoryPostUpgrade > currentDom0Memory)
+                {
+                    // add warning that the the Dom0 memory will be changed after the upgrade 
+                    return new HostMemoryPostUpgradeWarning(this, Host, dom0MemoryPostUpgrade);
+                }
+            }
+            else
+            {
+                // we don't know the Dom0 memory after the upgrade, so add generic warning if they are upgrading from pre-Naples 
+                if (Helpers.NaplesOrGreater(Host))
+                    return null;
+
+                string upgradePlatformVersion = null;
+                string upgradeProductVersion = null;
+                if (installMethodConfig != null)
+                    TryGetUpgradeVersion(out upgradePlatformVersion, out upgradeProductVersion);
+                
+                if (string.IsNullOrEmpty(upgradePlatformVersion))

```

Comment:
Wouldn't it be more generic and safer if we did
```
if (Helpers.NaplesOrGreater(upgradePlatformVersion))
{
    // we know that they are upgrading to Naples or greater, so add specific warning
    return new HostMemoryPostUpgradeWarning(this, Host, 0, upgradeProductVersion);
}
else
{
    // we don't know the upgrade version, so add generic warning (this is the case of the manual upgrade or when the rpu plugin doesn't have the function)
        return new HostMemoryPostUpgradeWarning(this, Host);
}
```
unless it's completely unlikely to have garbage returned in the json or we don't want to show any warning in this case.

Reply:
This was intended to also cover the case where we know the upgrade platform version and it is pre-Naples, in which case we shouldn't display a warning. But actually, we shouldn't get here in that case, due to lines 38-39